src/compilation/compiler.rs

```diff
@@ -243,88 +249,81 @@ impl Compiler {
+            false => LLVMSetLinkage(func, LLVMInternalLinkage),
+        }
+
         // TODO: Design: Remove variable shadowing
         //  By that, do you think we meant allow variables to shadow functions inside a scope?
         //  If so, I think we do that...? Not sure.
         let func_name = &func_proto.name;
-        // TODO: Minor: Remove clone (like accept owned through function?)
-        let func_proto = func_proto.clone();
-        self.scope_manager.set_func(func_name, func_proto, func);
+        self.scope_manager.set(func_name, func);
     }
 
     /// Complies a function definition, assuming the function's prototype has been compiled.
-    unsafe fn compile_func_def(&mut self, func_def: &FuncDef) {
+    unsafe fn compile_func_def(&mut self, func_def: &TypedFuncDef) {
         let func_name = CString::new(func_def.proto.name.as_str()).unwrap();
         let func = LLVMGetNamedFunction(self.module, func_name.as_ptr());
-        if func.is_null() {
+        if LLVMIsNull(func) == 1 {
             panic!(
                 "The prototype for function '{}' has not been defined",
                 func_def.proto.name
             );
         }
 
-        // TODO: QUICK: Should linkage be set in the compile_func_proto function?
-        //  YES! IT SHOULD! Let's move is_public to be a part of func_def.proto.
-        match func_def.is_public {
-            true => LLVMSetLinkage(func, LLVMExternalLinkage),
-            false => LLVMSetLinkage(func, LLVMInternalLinkage),
-        }
-
         let entry_block = LLVMAppendBasicBlockInContext(self.context, func, cstr!("entry"));
         LLVMPositionBuilderAtEnd(self.builder, entry_block);
 
         self.scope_manager.enter_scope();
 
         for (i, param) in func_def.proto.params.iter().enumerate() {
             let param_name = param.param_name.as_str();
-            let param_type = param.param_type;
+            let param_type = &param.param_type;
             let param_value_ref = LLVMGetParam(func, i as c_uint);
-            let alloca = self.create_entry_block_alloca(func, param_name, param.param_type);
+            let alloca = self.create_entry_block_alloca(func, param_name, param_type);
             LLVMBuildStore(self.builder, param_value_ref, alloca);
-            self.scope_manager.set_var(param_name, param_type, alloca);
+            self.scope_manager.set(param_name, alloca);
         }
 
         for statement in func_def.body.iter() {
             // TODO: Error if compiling statement fails?
             self.compile_statement(statement);
         }
 
-        // TODO: minor: Is this the best way of implicitly returning void?
-        //  It's correct. It might be redundant if the second-to-last statement
-        //  was a return. But I think this is fine.
-        // Implicitly return void if no return statement
-        match func_def.body.last() {
-            Some(Statement::Return(_)) => {}
-            _ => self.compile_ret_statement(&None),
-        }
-
```

Comment:
Good catch actually. The one thing is that our language right now seems to have a theme of making everything explicit. So the explicit thing to do is to require the user to put a return statement. That being said, I think I prefer auto-inserting a return for void functions because that will just get repetitive after a while in my opinion.

Also, yes it's always fine to insert a ret void at the end of a function because if it is redundant it will be optimized out. But even so, is the compiler optimizing it out something we want to rely on? Like it feels sorta like a hack.

If we do put it in, we should also check that the function's return type is None because that's the only type of function we can auto generate return statements for.

I honestly don't know whether I prefer leaving it in or taking it out. Do you have an opinion?

Reply:
Ok yes very good point! I think I'd rather have the return be optional in void functions, and I think having LLVM optimize it away is totally fine (LLVM is probably much better at detecting whether it's needed or not).

If we agree then tonight we can:
1. Put this back in
2. Panic if the function's return type isn't void (maybe something like `Non-void functions must end with a 'ret' statement` for now?)